SIMD: Increase Transaction Size Limit (#296)
solana-foundation/solana-improvement-documents: proposals/0296-increase-tx-size.md

can accomodate the max accounts used in ALTs directly in the transaction,
developers are able to migrate from `v0` transactions to `v1` transactions.

A number of changes are required to be made to the validator to support the new
transaction size limit. Consensus would need to be updated to support the
larger transaction size, as larger transactions included in a block would be
marked as invalid by the cluster today. The scheduler would also need to be
modified to  support the larger transaction sizes. While this proposal does not
introduce a new fee structure around bytes in a transaction, the scheduler
should prioritize larger transactions differently, requiring developers to pay
a higher priority fee to land their transaction.

Testing for this new transaction size limit should be done extensively to
ensure that performance on the cluster is not adversely affected. This testing
should include the different use cases mentioned within the motivation section,
as well as the different sizes of transactions that are currently being used by
developers on the network.

## Alternatives Considered

Alternative designs considered:

> 

Reviewer: I would like to add changes to update the `struct Packet` keeping the same size for 1280 for shreds gossip and turbine and a new struct for Transaction. Currently we use struct Packet and PACKET_DATA_SIZE everywhere which creates a lot of confusion. PACKET_DATA_SIZE should be left to 1232 while only for transaction size should be updated.

---
simd: '0296'
title: Increase Transaction Size
authors:
  - jacobcreech
  - apfitzge
category: Standard
type: Core
status: Review
created: 2025-05-28
feature: (fill in with feature key and github tracking issues once accepted)
---

## Summary

Historically, messages transmitted over the network must not exceed the IPc6 MTU
size to ensure a fast and reliable network delivery. Solana has used a 
conservative MTU size of 1280 bytes, which after accounting for some of the
overhead, leaves a maximum transaction size of 1032 bytes for data in the
payload.

With QUIC, we can now use a larger transaction size than the original 1280

> bytes. This proposal presents a new transaction size starting at 4096 bytes,

Reviewer 2: Since this transaction format aims to obviate the need for address lookup tables, and address lookup tables can today hold 256 32-byte addresses, isn't 4096 bytes too low to achieve the goal?

Author: 4096 bytes achieves the goal for now until max account load increases from 64. There's still the case where if you have an address be repeated many times over a ALT will be more efficient on packing today, but I don't see much usage like that in the wild.

and the new v1 transaction format required to support the larger size. v1 
transactions are a direct upgrade from the legacy format, while proposing 

> a removal of Address Lookup Tables compared to v0 transactions.

Reviewer 3: Can we also drop durable nonces in v1?
The reason to remove ALT is to be able to validate a tx without hitting the account db. Durable nonces have the same issue, we can't validate a tx without hitting the account db. Hence the proposal to drop them.

Author: I'm open to it though if there's a path forward

Author: v0 and v1 will coexist in the near future, but ideally we'd deprecate and remove v0. It'll be a long migration though and we'd need a solution to nonce transactions before we deprecate v0 and legacy.

Author: If we remove nonce support from the tx format, what else is needed to be done on v1 transactions? It seems like consensus from both Anza and FD is to remove in v1 so we have a migration path to remove nonces entirely in the future.

developers are using, but it does not have the same guarantees as a transaction
does on the protocol level, namely that bundles are not guaranteed to be atomic
like singular transactions.

For these reasons, increasing the current transaction size limit would enable
developers to create new applications and use cases that were previously
unfeasible.

## New Terminology

- 'v1 transaction' - A new transaction format that is designed to enable larger
  transactions sizes while not having the address lookup table features 
  introduced in v0 transactions.

## Detailed Design

Increasing the transaction size limit above the current MTU max size is
possible with the use of QUIC. QUIC's RFC 9000 specification does not have an
explicit maximum stream size, allowing for larger transactions to be sent.

A new transaction format, `v1`, is proposed to enable larger transaction sizes.
The `v1` transaction format would be:

```
VersionByte (u8) - >127 to distinguish from legacy/v0 formats
LegacyHeader (u8, u8, u8) -- Required signatures from the current
`MessageHeader` type
Payload Length (u16) -- Total size excluding signatures

> NumInstructions (u16)

Reviewer 4: - should this be our compact u16 encoding?
- can we put `NumInstructions` next to `Ixs` in the serialized format, to make encoding/decoding easier

Author: I tried to keep all the variable size fields after the static fields. In this case, only one of `NumAddresses` or `NumInstructions` can be next to their associated `Addresses` or `Ixs`

Reviewer 4: It's preferable in this case to have the size and the variable size array next to each other. The reason is that we don't treat these independently when we encode/decode.

When we decode, we first read the size, and then we use that to inform how many bytes we should read as the array. If they're not sequential then we need to read the size and then later use it to read the array. It complicates the process because we're carrying state forward that we could deal with in the same operation if they're sequential.

When we encode, we don't encode the size separately to the array. We encode to an array with a size prefix. If we can't do that then we end up with two separate encoding steps at different places in the sequence.

All of this can be dealt with if we need to obviously, but it's very standard to encode arrays with their size prefix. It will not be easier for us if you put all the static fields first. Obviously this convenience may be a lower priority if doing it this way is good for performance or something. But if there's no advantage to putting all the static fields first, then it would be preferable to keep size prefixed arrays whenever possible.

Reviewer 5: Reading compressed u16s is one of the most expensive parts of transaction parsing.
Out of band data effectively lets us skip parsing entirely, because the transactions header is effectively what the TransactionView is now - some static fields and then some offsets/indices into the raw packet. 

I did some experimentation with this last August when looking at tx ingest speed; it's not this exact proposal but involved removing compressed u16s and moving ix data out of band. For simple transfers (already very fast) it was ~30% faster. For many ixs (packed transfers) it was 5x faster.

Reviewer 4: > For simple transfers (already very fast) it was ~30% faster. For many ixs (packed transfers) it was 5x faster.

Got it. That's the context I was missing, thanks!

ResourceRequestMask (u16) -- Bitmask of which resource requests are present.
 For example bit 0 may mean "requested_cus", bit 1 may mean "microlamports per

>  cu"

Reviewer 4: Will this SIMD precisely define this?  Or will that be done when fees are figured out for v1 transactions? Great that this will reduce CU overhead so much!

Reviewer 6: I am not sure that this "resource bucket" design is the best way to tackle CU limits and prices. This looks like a premature optimisation that will become useless as soon as some new resource type needs more than a simple `u64` for storage. In fact, even with the current design, we're wasting 32 bits for CU limits (which are likely to be set on every single transaction) just because microlamports need 64 bits.

Since we have transaction versioning, I believe we should leverage this to add more resources in the future as and when we fully understand their spec. Otherwise, we'll likely end up with an ambiguous design where some resources that fit in a `u64` will be packed in v1 but if you need these other more complex resources you'll need to use v2+.

I suggest being explicit about the resources supported by v1 — i.e. CU limits and microlamports per CU — and let future versions worry about future resources. That way, we can have a more compact design such as:

```
ResourceRequestMask (u8) — mask[0] indicate that a CU limit is requested, mask[1] indicates that a CU price is requested, mask[2..] are unused.
ComputeUnitLimit (optional u32) — If and only if `mask[0] === 1`.
ComputeUnitPrice (optional u64) — If and only if `mask[1] === 1`.
```

Reviewer 4: We should probably include [loaded accounts data size](https://github.com/solana-program/compute-budget/blob/main/clients/js/src/generated/instructions/setLoadedAccountsDataSizeLimit.ts) too for this one

Reviewer 5: @jacobcreech we should specify some values here. Doesn't really matter which bit is what, but we should at least get rid of the compute unit request/price, acct data size. i.e. we should be able to deprecate ComputeBudgetProgram if everyone were to migrate to V1.

Author: Added proposed values that maps to everything the ComputeBudgetProgram has today.

NumStaticKeys (u8)

> RecentBlockhash (hash)

Reviewer 4: Are we removing durable nonce support for v1 transactions?

Reviewer 2: If not, then rename this to `LifetimeSpecifier`.

Reviewer 7: As part of these discussions, it was proposed to get rid of nonce because it hits accountsdb, see https://github.com/solana-foundation/solana-improvement-documents/pull/296/files#r2129100914

Reviewer 5: We can't let noncery block this proposal. Rename it to `LifetimeSpecficier` seems fine.

Author: I'm still in favor of removal of nonce support for v1 transactions. With v0 and legacy still coexisting with v1, this shouldn't be a blocker.

> StaticKeys [[u8; 32]] -- Number matches NumStaticKeys

Reviewer 4: Similar, can we put `NumStaticKeys` and `StaticKeys` next to each other? Makes en/decoding easier

Reviewer 2: Because this proposal would eliminate address lookup tables, aren't all keys ‘static’ now?

Propose renaming this from `StaticKeys` to `Addresses` since that's what they _really_ are. We've been fighting this ‘address vs. public key’ impedance mismatch for a while (eg. program derived addresses are not public keys, in any useful sense of the word) and this SIMD is as good a time as any to fix it.

Author: Updated. I agree with this sentiment.

ResourceRequests [u64] -- Array of request values. (section size is popcount
 ResourceRequestMask * 8). each value is a u64.

> Ixs [(u16, u16, u16, u16)] -- Number matches NumInstructions. Values are

Reviewer 2: Why is `num_accounts` a u16 if the upper limit of ‘static keys’ above is u8? Without knowing what the format of an address is in `TRAILING_DATA_SECTION` it's hard to comment.

Reviewer 5: It's a compressed u16 now in the serialization format. Just went with a u16 here to avoid having to change it.

Accounts can be repeated. I'm in favor of just changing it to a `u8`, since we need to also add the program_index in for each of `Ixs`:

@jacobcreech We should do:

```
Ixs [(u8, u8, u16, u16, u16)] -- Number matches NumInstructions. Values are
 (program_account_index, num_accounts, accounts_offset, num_data_bytes, bytes_offset)
```

>  (num_accounts, accounts_offset, num_data_bytes, bytes_offset).

Reviewer 5: It's unclear to me where you're getting 2 extra bytes per account. The design is definitely less space efficient, that's fine from our view since it's paired with a tx-size increase. It's better have constant sized fields than deal with compressed u16s inherent branchiness.

> [TRAILING DATA SECTION] -- Length is such that PayloadLength is matched.

Reviewer 4: I don't understand exactly how we're encoding instructions here, or why we're changing it.

First of all to clarify, is `TRAILING_DATA_SECTION` entirely for instructions?

Currently for each instruction we have:

- programAddressIndex (u8)
- accountIndices (array of u8, compact-u16 size)
- data (array of u8 bytes, compact-u16 size)

These are sequential, so we don't need to encode offsets.

Based on the offsets specified, I assume `TRAILING DATA SECTION` is intended to encode `accountIndices` and `data` for all instructions?

How is `programAddressIndex` being encoded? Is it the first account index for each instruction?

In general, why are we changing how we encode instructions? I don't see any advantage and I think this will end up using more bytes, but I might be missing something.

Reviewer 4: For the current proposal, I think the `Ixs` types should be `u8, u8, compact-u16, compact-u16`

Reviewer 6: 100% agree!

I wasn't sure if `u16`s were explicitly replacing `compact-u16`s or if it was just omitted by mistake. If it's the former, it would be nice to make it clear and explain why we are making this change.

I also agree that the new instruction layout seems odd and doesn't seem to add anything to the current layout unless I'm missing something.

My guess is that the change from `u8` to `u16` in account indices is to support more than 256 accounts per transaction. Unless you're using Address Lookup Tables though, `32 * 256 = 8kb` which is double the new limit anyways. Perhaps this change is still not worth it since it would only be optimising for heavy usage of Address Lookup Tables whereas it will double the size of indices everywhere else (which are often used multiple time per account).

Reviewer 5: compressed u16s are horrible for performance. they are removed entirely.

> My guess is that the change from u8 to u16 in account indices.

It doesn't do this though, `u8` is still used for the account indices themselves.

Reviewer 5: > How is programAddressIndex being encoded? Is it the first account index for each instruction?

I didn't include it in my drafted up notes before I left, so it isnt' in the proposal that Jacob posted. We'd want a `u8` in each of the `Ixs`. Need to fix.

Reviewer 4: Thanks for the context on the performance @apfitzge! I think my only remaining concern is just having a really clear spec of exactly how `TRAILING DATA SECTION` is to be structured.

Reviewer 5: I hope this comment (and the soon modified SIMD) will clear it up! https://github.com/solana-foundation/solana-improvement-documents/pull/296/files#r2152648212

migrate to the new transaction format. Similarly, since the new limit proposed
can accomodate the max accounts used in ALTs directly in the transaction,
developers are able to migrate from `v0` transactions to `v1` transactions.

A number of changes are required to be made to the validator to support the new
transaction size limit. Consensus would need to be updated to support the
larger transaction size, as larger transactions included in a block would be
marked as invalid by the cluster today. The scheduler would also need to be
modified to  support the larger transaction sizes. While this proposal does not
introduce a new fee structure around bytes in a transaction, the scheduler
should prioritize larger transactions differently, requiring developers to pay
a higher priority fee to land their transaction.

Testing for this new transaction size limit should be done extensively to
ensure that performance on the cluster is not adversely affected. This testing
should include the different use cases mentioned within the motivation section,
as well as the different sizes of transactions that are currently being used by
developers on the network.

> 

Reviewer 3: We should add a section on explicit constraints.
In v0, most constraints are implicitly based on the 1232 magic number, for example a tx can only have 12 signatures -- but this makes validation a bit fuzzy and in general these implicit constraints are just a worst case that we have to handle but never happens.

Explicit constraints make validation more stable across implementation, and will let us increase them independently. For example we can increase the total size of the tx and/or the max size of an instruction data, without necessarily increasing the max number of instructions or accounts (allow to support larger signatures/zk proofs, but without repercussion on number of ixs that the runtime has to process).

Here's an initial list:

| max | value | comment |
| --- | --- | --- |
| max tx size | 4096 | |
| max num signatures | 8 | same as precompiles |
| max num accounts | 64 | with feature gate to increase to 128 |
| max num instructions | 32 | |
| max accounts/instruction | 128 | accounts can be repeated |
| max data bytes/instruction | 3900 | |
| max UDP packets (fragmentation) | 6 | not for consensus, but prob useful to spec |

For num instructions, I proposed 32 as the use case I can think of is for exchanges to do many transfers in a single tx (useful on eth/btc, less useful on sol?). If you have 3-4 accounts/ix and a max of 64 accounts, 32 ixs should be more than reasonable (but can do more, no strong opinion).

For max data/instruction, I just put some sort of worst case. I'd personally lower it a bit, but don't want to start a war. :)

For max UDP packets, Agave currently limits to 4 (for v0=1232), though the actual max that it produces is 2 (quinn, configured with `send_fairness(false)`). Based on QUIC specs, one could even send a 1232B tx in 1232 UDP packets.

Reviewer 2: Interesting fact regarding being able to pack more addresses in a transaction. [SIMD-163](https://github.com/solana-foundation/solana-improvement-documents/pull/163) will make it so that an instruction no longer has to include a program's address in its list (eg. the system program) to be able to CPI to it (eg. make a system_program::transfer). Presence of the program address in the top level account list will be sufficient.

This is to say that there will be downward pressure on how many addresses an instruction needs to include to be able to run, just as an FYI when making decisions about limits.

Author: @0x0ece Are the new limits in your list max tx size, max data bytes/instruction, and max UDP packets? I believe all the others are currently set elsewhere in previous features.

Reviewer 3: @jacobcreech Currently many of these limits are **implicit**, someone computed them based on the 1232 tx size (e.g. max signatures is 12). But if we increase the size it's unclear what to do for each limit (e.g. max signatures remains 12, grows and if so to what?). So all I'm saying is to just have a simple table with all of them **explicit**.

Reviewer 8: What about block-level constraints? The block size is implicitly constrained by the CU limit so, if we keep the base CUs per tx constant, the maximum block size can theoretically increase nearly 4x.

Rather than increasing the base CUs, I'd support an explicit block size constraint and have tx size removed from consideration for the purposes of CU tracking. This has additional benefits, like allowing increases to the block CU limit without needing to support larger blocks.

Reviewer 9: just walking in here to veto the max instruction limit. there's already an explicit limit in the runtime (currently max 64) and we should just make it equal to that

Reviewer 10: imo:

good? |max | value | comment
-- | -- | -- | --
✅ | max tx size | 4096 |  
❌ | max num signatures | 8 | same as precompiles
✅ | max num accounts | 64 | with feature gate to increase to 128
❌ | max num instructions | 32 |  
✅ | max accounts/instruction | 128 | accounts can be repeated
❌ | max data bytes/instruction | 3900 |  
❌ | max UDP packets (fragmentation) | 6 | not for consensus, but prob useful to spec

In other words, we probably only have to rigidly define max tx size, as the others that matter are already explicitly defined. We should not stop people from sending nodes more lamports to verify more signatures. We also should not be opinionated on how people choose to handle their transport for transaction broadcast. I should be able to send it to you by carrier pigeon if I want.

Author: Over the past 7 days we've seen instruction counts as high as 52 and a max of 10 signatures on transactions. I wouldn't want to break those with lower limits.

Reviewer 3: Great, I'm fine with any limit. I just want to know it explicitly, not have to infer it from Agave's code.

Let's keep 12 signatures (or even raise it to 63 if you think it's any useful) and max ixs definitely to 64 (I wasn't aware of that, it wasn't an attempt to lower it).

Reviewer 11: To be clear, we're not asking for any _new_ limits, just documenting in this SIMD the ones that are there that will be affected by this change.

Author: Added a table addressing these exact limitations

## Alternatives Considered

Alternative designs considered:

- Having a transaction loading feature that would allow developers to load
the transaction parts in a buffer and then be able to execute them at the
end.
This method is no longer considered as it requires a much higher level of
latency on the application level and a decent amount of complexity within
the validator
- Bundles at the protocol level. This would not solve all problems that are
 solved by larger transaction sizes. Bundles would still limit the ability for
 developers to use cryptographic signature schemes that have large proof sizes.

## Impact

Developers would have to update with applications to use the new transaction
format to take advantage of the larger transaction size. Those developers that
had previously been using address lookup tables would be required to update the
new transactions with the full address list instead of the address lookup table
and its indices.

Over the long term, the new transaction format should require a migration to
fully deprecate both `legacy` and `v0` transaction formats, and in deprecating
`v0` transactions, the address lookup table feature would be removed from the

> protocol.

Reviewer 2: Recommend removing this.

Deprecations like this are tractable when we can plausibly expect the elimination of old consumers from the ecosystem (as is the case with the natural drawdown of old validator versions) but the continued existence of dApps that rely on the `legacy` and `v0` transaction formats to function is assured. To remove support for a transaction format is to break every application that relies on those formats. The highlighted impact statement doesn't say that outright.

---
simd: '0296'
title: Increase Transaction Size
authors:
  - jacobcreech
  - apfitzge
category: Standard
type: Core
status: Review
created: 2025-05-28
feature: (fill in with feature key and github tracking issues once accepted)
---

## Summary

Historically, messages transmitted over the network must not exceed the IPc6 MTU
size to ensure a fast and reliable network delivery. Solana has used a 
conservative MTU size of 1280 bytes, which after accounting for some of the
overhead, leaves a maximum transaction size of 1232 bytes for data in the
payload.

With QUIC, we can now use a larger transaction size than the original 1280

> bytes. This proposal presents a new transaction size starting at 4096 bytes,

Reviewer 12: we will likely support the increased tx size in bundles

client as well as a frictional developer experience for many developers.

A number of developers have also resorted to using Jito bundles to artificially
increase the size of their transactions. This is a current workaround that
developers are using, but it does not have the same guarantees as a transaction
does on the protocol level, namely that bundles are not guaranteed to be atomic
like singular transactions.

For these reasons, increasing the current transaction size limit would enable
developers to create new applications and use cases that were previously
unfeasible.

## New Terminology

- 'v1 transaction' - A new transaction format that is designed to enable larger
  transactions sizes while not having the address lookup table features 
  introduced in v0 transactions.

## Detailed Design

Increasing the transaction size limit above the current MTU max size is
possible with the use of QUIC. QUIC's RFC 9000 specification does not have an
explicit maximum stream size, allowing for larger transactions to be sent.

A new transaction format, `v1`, is proposed to enable larger transaction sizes.
The `v1` transaction format would be:

```

> VersionByte (u8) - >129 to distinguish from legacy/v0 formats

Reviewer 11: Sure, the idea was less about supporting huge numbers of different transaction formats and more about having a static magic value so a hardware parser at the edge could easily determine the type/detect corruption early etc.

LegacyHeader (u8, u8, u8) -- Required signatures from the current
`MessageHeader` type

> Payload Length (u16) -- Total size excluding signatures

Reviewer 13: I'm not sure the payload length is adding much value. Can someone remind me why this is desired? cc @apfitzge 

It seems like it will sanitization slightly more involved because rather than just assuming the message is the rest of the packet data, we have to sanitize this payload length to make sure it's long enough for the decoded message and that the message doesn't overlap with the signatures.

> NumInstructions (u8)

Reviewer 13: Noting that representing `NumInstructions` as a `u8` is a new restriction. Previously, I believe transactions could have over 255 instructions but the SVM limits execution such that only the first 64 instructions (including CPI's) are executed.

> ResourceRequestMask (u64) -- Bitmask of which resource requests are present.

Reviewer 5: Let's rename `ResourceRequestMask` to something a bit more generic like `TransactionConfigMask`.

priority fee/cu price is not really a "resource" request.

 - 0: requested_compute_unit_limit
 - 1: requested_compute_unit_price
 - 2: requested_heap_size

>  - 3: requested_loaded_accounts_data_size_limit

Reviewer 13: Continuing thread here: https://github.com/solana-foundation/solana-improvement-documents/pull/296#discussion_r2317620952

LifetimeSpecificier [u8; 32]
NumAddresses (u8)

> Addresses [[u8; 32]] -- Number matches NumAddresses

Reviewer 13: nit: can you change "Number" to "Length" here and for `Ixs`?

ConfigRequests [u64] -- Array of request values. (section size is popcount
 TransactionConfigMask * 8). each value is a u64.
Ixs [(u8, u8, u16, u16, u16)] -- Number matches NumInstructions. Values are 
 (program_account_index, num_accounts, accounts_offset, num_data_bytes, 
 bytes_offset)
[TRAILING DATA SECTION] -- Length is such that PayloadLength is matched.
Signatures [[u8; 64]] -- Length of `num_required_signatures` from
 `LegacyHeader`
```

The [TRAILING DATA SECTION] contains instruction's serialized account indexes

> and instruction data bytes,

Reviewer 5: for performance. variably sized data is bad for speed. w/ this structure all the ix meta is stored in a simple array, something like iterating the list to look at the u8 program_index (not reading ix data or ix accts) would be much faster this way.

It also speeds up initial validation - i can loop the ix meta and sum up the lengths. never need to read the data/accts, just check total lenght matches the expectation.

in the order they appear in the transaction. Each instruction's account indexes
must appear before its data bytes.
For example, if there are two instructions:

1. `Instruction { account_indexes: [0, 1], data: [2, 3, 4] }`
2. `Instruction { account_indexes: [5, 6], data: [7, 8] }`

The TRAILING DATA SECTION must look like this:

```
[0, 1, 5, 6, 2, 3, 4, 7, 8]
```

Lengths are serialized in the earlier `Ixs` section, and so are not repeated in
the trailing data section.

This new `v1` transaction format notably does not include address lookup
tables.

> In consideration of what size the new transaction size limit should increase

Reviewer 13: Everything from this line downwards shouldn't be in the details section IMO. The details should only include information needed for validator clients to implement the proposal.

the validator
- Bundles at the protocol level. This would not solve all problems that are
 solved by larger transaction sizes. Bundles would still limit the ability for
 developers to use cryptographic signature schemes that have large proof sizes.

## Impact

Developers would have to update with applications to use the new transaction
format to take advantage of the larger transaction size. Those developers that
had previously been using address lookup tables would be required to update the
new transactions with the full address list instead of the address lookup table
and its indices.

## Security Considerations

Larger transaction sizes could results in a number of bandwidth issues on the 
cluster than need to be tested thoroughly to ensure that performance is not
adversely affected.

## Drawbacks 

There is an argument to be made that an increase in the transaction size limit
above the max MTU packet size would incur some network overhead due to
fragmentation retransmits. For a validator, this could mean maintaining
in-memory buffers for larger transactions compared to just receiving and
processing a single UDP packet. These issues will need to be addressed in the

> scheduler and how priority fees for larger transactions are handled.

Reviewer 11: Agree - sorry my comment was maybe not clear - I was not asking to amend this SIMD to address this, just commenting on a potential long-term downside. But I think we've all agreed we should go ahead with this change, without synchronizing network- and application-fragmentation, for the reasons you state.

client as well as a frictional developer experience for many developers.

A number of developers have also resorted to using Jito bundles to artificially
increase the size of their transactions. This is a current workaround that
developers are using, but it does not have the same guarantees as a transaction
does on the protocol level, namely that bundles are not guaranteed to be atomic
like singular transactions.

For these reasons, increasing the current transaction size limit would enable
developers to create new applications and use cases that were previously
unfeasible.

## New Terminology

- 'v1 transaction' - A new transaction format that is designed to enable larger
  transactions sizes while not having the address lookup table features 
  introduced in v0 transactions.

## Detailed Design

Increasing the transaction size limit above the current MTU max size is
possible with the use of QUIC. QUIC's RFC 9000 specification does not have an
explicit maximum stream size, allowing for larger transactions to be sent.

A new transaction format, `v1`, is proposed to enable larger transaction sizes.
The `v1` transaction format would be:

> ```

Reviewer 14: I think we need to make this more detailed and spec-like. 

```
VersionByte (u8)
LegacyHeader (u8, u8, u8) 
NumInstructions (u8)
TransactionConfigMask (u32) -- Bitmask of which config requests are present.
LifetimeSpecifier [u8; 32]
NumAddresses (u8)
Addresses [[u8; 32]] -- Length matches NumAddresses
ConfigValues [[u8; 4]] -- Length equal to the popcount (number of set bits) of
  TransactionConfigMask. See section TransactionConfigMask for details.
InstructionHeaders [(u8, u8, u16)] -- Length of NumInstructions. Values are 
  (ProgramAccountIndex, NumInstructionAccounts, NumInstructionDataBytes)
InstructionPayloads [InstructionPayload] -- Length = NumInstructions.
  Each InstructionPayload is the concatenation of the following byte arrays:
    InstructionAccountIndexes [u8] -- Length = NumInstructionAccounts from the 
    corresponding InstructionHeader
    InstructionData [u8] -- Length = NumInstructionDataBytes from the 
    corresponding InstructionHeader
Signatures [[u8; 64]]
```

#### VersionByte
The VersionByte MUST be set to `129` to distinguish v1 transactions from legacy/v0 formats.

#### LegacyHeader
The LegacyHeader is unchanged from prior transaction formats.  
For reference, it consists of three `u8` values with no padding.  The values (in order) are:
*   `num_required_signatures`: The number of requires signatures
*   `num_readonly_signed_accounts`: The number of the accounts with signatures that are loaded as read-only
*   `num_readonly_unsigned_accounts`: The number of accounts without signatures that are loaded as read-only

As in prior formats, the transaction fails sanitization if `num_readonly_signed_accounts >= num_required_signatures`.  Note that `num_readonly_signed_accounts == num_required_signatures` is a sanitization failure because it would imply the fee payer account is readonly.  Note that this also implies `num_required_signatures >= 1`.

#### NumInstructions
NumInstructions is set to the number of instructions in the transaction.  As explained in the Transaction Constraints section, the transaction fails sanitization if `NumInstructions > 64`.

#### TransactionConfigMask
TransactionConfigMask is explained below in detail.

#### LifetimeSpecifier
LifetimeSpecifier is the same as the Recent Blockhash field in prior transaction formats. It has been renamed to clarify its use without changing its meaning.

#### NumAddresses
NumAddresses is set to the number of account addresses the transaction references.  The transaction fails sanitization if `num_addresses < num_required_signatures + num_readonly_unsigned_accounts`. The transaction also fails sanitization if `num_addresses > 64`, as explained in the Transaction Constraints section.

#### Addresses
Addresses is an array of all the account addresses that the transaction references.  It is a sanitization failure for this array to contain any duplicates.  This list has `NumAddresses` elements.  The ordering of the addresses is unchanged from prior transaction formats.
For reference, they are:
* `num_required_signatures-num_readonly_signed_accounts` additional addresses for which the transaction contains signatures and are loaded as writable, of which the first is the fee payer
* `num_readonly_signed_accounts` addresses for which the transaction contains signatures and are loaded as readonly
* `num_addresses-num_required_signatures-num_readonly_unsigned_accounts` addresses for which the transaction does not contain signatures and are loaded as writable
* `num_readonly_unsigned_accounts` addresses for which the transaction does not contain signatures and are loaded as readonly

Any section with 0 addresses is skipped.

#### ConfigValues
ConfigValues is explained below with TransactionConfigMask in detail.

#### InstructionHeaders
InstructionHeaders is an array with NumInstructions elements.  Each element consists of three fields with no padding:
* `ProgramAccountIndex: u8`: the index in the array of addresses of the address of the program to invoke for this instruction.  This field was known as `program_id_index` in previous versions of the transaction format; it has been renamed to clarify its use without changing its meaning.
* `NumInstructionAccounts: u8`: the number of addresses that will be passed to the program when it is invoked.  This field was implicitly represented in a different encoding in a previous versions of the transaction format as the element count of the `accounts` vector.
* `NumInstructionDataBytes: u16`: the size in bytes of the data that will be passed as input to the program when it is invoked.  This field was implicitly represented in a different encoding in a previous versions of the transaction format as the element count of the `data` vector.

There is also no padding between each 4-byte element.

#### InstructionPayloads
`InstructionPayloads` is a concatenation of byte arrays.  For each of the `NumInstructions` instructions, it consists of `NumInstructionAccounts` bytes of account indices followed by `NumInstructionDataBytes` of instruction data.  For example, the InstructionPayloads for a two-instruction transaction looks like 
* instruction 0's InstructionAccountIndices (InstructionHeaders[0].NumInstructionAccounts bytes)
* instruction 0's InstructionData (InstructionHeaders[0].NumInstructionDataBytes bytes)
* instruction 1's InstructionAccountIndices (InstructionHeaders[1].NumInstructionAccounts bytes)
* instruction 1's InstructionData (InstructionHeaders[1].NumInstructionDataBytes bytes)

The transaction fails sanitization if any element of InstructionAccountIndices is greater than or equal to NumAddresses

#### Signatures
Signatures is an array with `num_required_signatures` elements, each consisting of a 64-byte Ed25519 signature.  Specifically, `Signatures[i]` is the signature resulting from signing the portion of the transaction prior to the Signatures field with the keypair associated with the public key at `Addresses[i]`.

The transaction must not contain trailing data after the signatures field.  There are no padding fields.

Author: I like this breakdown. I've updated

LifetimeSpecifier [u8; 32]
NumAddresses (u8)
Addresses [[u8; 32]] -- Length matches NumAddresses
ConfigValues [[u8; 4]] -- Length equal to the popcount (number of set bits) of
  TransactionConfigMask. See section TransactionConfigMask for details.
InstructionHeaders [(u8, u8, u16)] -- Length of NumInstructions. Values are 
  (ProgramAccountIndex, NumInstructionAccounts, NumInstructionDataBytes)
InstructionPayloads [InstructionPayload] -- Length = NumInstructions.
  Each InstructionPayload is the concatenation of the following byte arrays:
    InstructionAccountIndexes [u8] -- Length = NumInstructionAccounts from the 
    corresponding InstructionHeader
    InstructionData [u8] -- Length = NumInstructionDataBytes from the 
    corresponding InstructionHeader
Signatures [[u8; 64]] -- Length of `num_required_signatures` from
 `LegacyHeader`
```

This new `v1` transaction format notably does not include address lookup
tables.

### Transaction Constraints

Any transaction violating these constraints will be considered invalid and will 
not be included in the chain. Violations are considered sanitization failures:

> | max | value | new/existing |
> | --- | --- | --- |
> | max transaction size | 4096 | new |
> | max num signatures per transaction | 42 | new |
> | max num accounts | 96 | new |
> | max num instructions | 64 | old |
> | max accounts/instruction | 255 | new |
> | max num signatures per transaction | 12 | new |
> | max num accounts | 64 | old |
> | max accounts/instruction | 255 | old |

Reviewer 14: ```suggestion
| value | max specified by this SIMD | prior max | potential future max |
| --- | --- | --- | --- |
| transaction size | 4096 | 1232 | 4096 |
| signatures per transaction | 12 | 12 | 42 |
| num accounts | 64 | 64 | 96 |
| num instructions | 64 | 64 | unspecified |
| accounts/instruction | 255 | 255 | 255 |
```

The `prior max` and `potential future max` columns are for informational purposes only and are not a normative part of this specification.

Reviewer 14: As is, this table doesn't make much sense. It's not clear how to understand the duplicate rows.

Reviewer 14: @jstarry You were asking for the last column, right? I'm not sure exactly what you were thinking. What about this?

Author: This makes sense to me. I've updated to match

  (ProgramAccountIndex, NumInstructionAccounts, NumInstructionDataBytes)
InstructionPayloads [InstructionPayload] -- Length = NumInstructions.
  Each InstructionPayload is the concatenation of the following byte arrays:
    InstructionAccountIndexes [u8] -- Length = NumInstructionAccounts from the 
    corresponding InstructionHeader
    InstructionData [u8] -- Length = NumInstructionDataBytes from the 
    corresponding InstructionHeader
Signatures [[u8; 64]] -- Length of `num_required_signatures` from
 `LegacyHeader`
```

This new `v1` transaction format notably does not include address lookup
tables.

### Transaction Constraints

Any transaction violating these constraints will be considered invalid and will 
not be included in the chain. Violations are considered sanitization failures:

| max | value | new/existing |
| --- | --- | --- |
| max transaction size | 4096 | new |
| max num signatures per transaction | 42 | new |
| max num accounts | 96 | new |
| max num instructions | 64 | old |
| max accounts/instruction | 255 | new |

While the above lists sanitization failures, we will add a runtime max limit of 

> 12 signatures per transaction to be consistent with the current implicit limit.

Author: Added a table for Runtime Contraints

  TransactionConfigMask. See section TransactionConfigMask for details.
InstructionHeaders [(u8, u8, u16)] -- Length of NumInstructions. Values are 
  (ProgramAccountIndex, NumInstructionAccounts, NumInstructionDataBytes)
InstructionPayloads [InstructionPayload] -- Length = NumInstructions.
  Each InstructionPayload is the concatenation of the following byte arrays:
    InstructionAccountIndexes [u8] -- Length = NumInstructionAccounts from the 
    corresponding InstructionHeader
    InstructionData [u8] -- Length = NumInstructionDataBytes from the 
    corresponding InstructionHeader
Signatures [[u8; 64]] -- Length of `num_required_signatures` from
 `LegacyHeader`
```

This new `v1` transaction format notably does not include address lookup
tables.

### Transaction Constraints

Any transaction violating these constraints will be considered invalid and will 
not be included in the chain. Violations are considered sanitization failures:

| max | value | new/existing |
| --- | --- | --- |
| max transaction size | 4096 | new |
| max num signatures per transaction | 42 | new |
| max num accounts | 96 | new |
| max num instructions | 64 | old |
| max accounts/instruction | 255 | new |

> ### Runtime Contraints

Reviewer 13: ```suggestion
### Runtime Constraints
```

> ### Runtime Contraints

Reviewer 13: @ptaffet-jump I hadn't considered those edge cases. I agree they should all be static protocol constraints that are agreed to potentially be increased in future feature flags, that's a much better approach.

My only concern was avoiding wallets signing messages that make assumptions about what could never be a valid transaction in the future.

> ### Runtime Contraints

Author: I moved all of the runtime contraints and made them as part of the santization contraints.

While the above lists sanitization failures, we will have the following runtime
contraints either stay unchanged by making sure they are explicitly defined:

| max | value | new/existing |
| --- | --- | --- |
| max num signatures per transaction | 12 | new |
| max num accounts | 64 | old |
| max accounts/instruction | 255 | old |

The runtime check for max number of signatures should be run immediately after

> fee payer validations and before other runtime checks.

Reviewer 13: ```suggestion
fee payer validations and before other runtime checks. If the number of
transaction signatures exceeds the max, return `TooManySignatures`.
```

> ### TransactionConfigMask

Reviewer 14: There's something to be said for this, but the downside is that the validator needs to check this for every transaction, in both the leader pipeline and in replay.

And while I'm on a role (https://github.com/solana-foundation/solana-improvement-documents/pull/296#discussion_r2383261425) if the only violation of the sorting order happens in accounts > 64, is that still a sanitization failure?

> ### TransactionConfigMask

Reviewer 13: The fee payer would have to be exempted from the sorting.

And what's the benefit of the following?

> This means that two parties who agree on the components of a transaction will always, deterministically, produce the same compiled transaction message.

I don't really see much value in this and the overhead for the validator is non trivial

> ### TransactionConfigMask

Reviewer 2: > The fee payer would have to be exempted from the sorting.

Good call. Updated. This is another thing that's [missing from the existing docs](https://solana.com/docs/core/transactions#array-of-account-addresses)!

> And what's the benefit of the following?
>
> > This means that two parties who agree on the components of a transaction will always, deterministically, produce the same compiled transaction message.

You and I, speaking through a telephone and _without_ sharing the compiled transaction bytes, can agree to propose a transaction that both of us must sign, without the risk of signature mismatches owing to which order we prepare the account list. The day that I deployed https://github.com/solana-foundation/solana-web3.js/commit/f119e024c17d15942bb7b4a531c75a4d7bc38650, a related problem arose between two parties – an application and a wallet. IIRC, upon receipt of a partially signed transaction from an application, the wallet decoded it for analysis, then _re-encoded_ it for signing. Because the application had sorted the account addresses one way, and the wallet sorted them a _different_ way upon re-encoding, the two signatures were no longer of the same message. Formalizing the account order in the spec will make this impossible.

Verifying the order will require ((96-1)x32 = **3040**) integer comparisons in the worst case, given 96 addresses and a failure to bail out until the last byte of each pair. I presume that verifying one Ed25519 signature is a couple of orders of magnitude more expensive, per signature.

> ### TransactionConfigMask

Reviewer 14: I think @apfitzge had prototyped the parser for the instruction format and it was ~1us. Perhaps he can add this check to the prototype and we can get a better sense of the overhead before we decide this.

> ### TransactionConfigMask

Author: > We don't control the entire ecosystem of libraries. As soon as one implementation disagrees on the ordering for lack of a specification in the protocol, the opportunity for a signature mismatch presents.

Just spec it a specific way in a SRFC and then the SDKs have to agree otherwise they're wrong. We know all of the SDK maintainers and have the ability to make application-specific specs. This way we can keep it out of the protocol, keep the speed, and just require SDK maintainers to adhere to the spec.

The transaction config mask is used to configure specific fee and resource 
requests in a transaction.
It is intended to be flexible enough to add new fields in the future, and is 
initially aimed at replacing the ComputeBudgetProgram instructions that are 
currently used to configure transactions.
Each bit in the mask represents 4-bytes in the `ConfigRequests` array.
If a configured value, such as priority-fee, needs more than 4-bytes a field can
use 2 bits in the mask.

Initially supported fields and assigned bits:

> - [0, 1] - priority-fee. 8-byte LE u64.

Reviewer 5: @jacobcreech could you just add a short "Total lamports for transaction priority-fee" here

client as well as a frictional developer experience for many developers.

A number of developers have also resorted to using Jito bundles to artificially
increase the size of their transactions. This is a current workaround that
developers are using, but it does not have the same guarantees as a transaction
does on the protocol level, namely that bundles are not guaranteed to be atomic
like singular transactions.

For these reasons, increasing the current transaction size limit would enable
developers to create new applications and use cases that were previously
unfeasible.

## New Terminology

- 'v1 transaction' - A new transaction format that is designed to enable larger
  transactions sizes while not having the address lookup table features 
  introduced in v0 transactions.

## Detailed Design

Increasing the transaction size limit above the current MTU max size is
possible with the use of QUIC. QUIC's RFC 9000 specification does not have an
explicit maximum stream size, allowing for larger transactions to be sent.

A new transaction format, `v1`, is proposed to enable larger transaction sizes.
The `v1` transaction format would be:

```

> VersionByte (u8) - >129 to distinguish from legacy/v0 formats

Reviewer 5: This should be =129 for v1. 130 would be v2, and so on.

LegacyHeader (u8, u8, u8) -- Required signatures from the current
`MessageHeader` type
NumInstructions (u8)
TransactionConfigMask (u32) -- Bitmask of which config requests are present.
LifetimeSpecificier [u8; 32]
NumAddresses (u8)
Addresses [[u8; 32]] -- Length matches NumAddresses

> ConfigRequests [u8] -- Array of request values. (section size is popcount 
>   TransactionConfigMask * 4). See section TransactionConfigMask for details.

Reviewer 14: To make this slightly more clear:
```suggestion
ConfigRequests [u8] -- Array of request values. (section size is 4*popcount( 
  TransactionConfigMask) bytes). See section TransactionConfigMask for details.
```

Ixs [(u8, u8, u16)] -- Number matches NumInstructions. Values are 
 (program_account_index, num_accounts, num_data_bytes)
[TRAILING DATA SECTION] -- Length must be consistent with the sum of 
 num_accounts and num_data_bytes in `Ixs`
Signatures [[u8; 64]] -- Length of `num_required_signatures` from
 `LegacyHeader`
```

The [TRAILING DATA SECTION] contains instruction's serialized account indexes
and instruction data bytes,
in the order they appear in the transaction. Each instruction's account indexes
must appear before its data bytes.
For example, if there are two instructions:

> 1. `Instruction { account_indexes: [0, 1], data: [2, 3, 4] }`

Reviewer 5: Agree, good to clarify even if the behavior is unchanged from the current format.

2. `Instruction { account_indexes: [5, 6], data: [7, 8] }`

The TRAILING DATA SECTION must look like this:

```

> [0, 1, 5, 6, 2, 3, 4, 7, 8]

Reviewer 5: @jacobcreech this should be

```
[0, 1, 2, 3, 4, 5, 6, 7, 8]
```

Reviewer 5: The instruction's trailing section is serialized per instruction:
[ix1 accts][ix1 data]
[ix2 accts][ix2 data]
...

Reviewer 15: How will the different instruction boundaries be computed? 
i.e. how will you know that `[5,6]` is the start of a new instruction? 

Might be performant,  but seems to be a worse data experience.

```

Lengths are serialized in the earlier `Ixs` section, and so are not repeated in
the trailing data section.

This new `v1` transaction format notably does not include address lookup
tables.

With the new transaction size limit, it is possible to go past some other 
implicit limitations that are currently in place on the network. To avoid
causing unintended issues, the following limitations are proposed to be in
place:

| max | value |
| --- | --- |
| max transaction size | 4096 |

> | max num signatures per transaction | 63 |

Reviewer 5: 63 signatures is not possible in a validly serialized packet.
Each signature requires a unique pubkey, so each signature will require at 96 bytes.

For a tx with no tx config, no other pubkeys, no instructions, the non-sig non-pubkey byte overhead will be 42 bytes. Leaving us with 4054 bytes, and allowing for a maximum (valid) number of signatures as 42 (rounding down to nearest integer).

> | max num signatures per transaction | 42 |

Reviewer 5: I don't see the reason why this should be restricted further than is permissible by the format.
The other limits are there because there are runtime restrictions related to them i.e. we do not allow >64 accts, nor >64 ixs.

> | max num signatures per transaction | 42 |

Reviewer 13: Yeah good point, I can be ok with a sanitization limit of something high like 42, but this SIMD should also define a separate runtime limit of like 12

> | max num signatures per transaction | 42 |

Reviewer 11: In general, is there a good reason for sanitization limits to be different than runtime limits? We've ran into some issues in the past where sanitization limits are higher than runtime limits, which makes it difficult to statically assign memory for parsed transactions and still be conformant (no bank hash mismatches). It's much easier to design a conformant client if the sanitization limits are the same as the runtime limits.

If the motivation for runtime/sanitization limits being different is that we may increase these limits in the future, could we just feature-gate the sanitization limit increase at the same time?

> | max num signatures per transaction | 42 |

Reviewer 11: One downside to making these higher - than we anticipate increasing the runtime limits to - is that fd has to allocate memory for worst-case parsed transaction sizes. Having these limits high increases our memory footprint and decreases throughput. So it helps us if the sanitization limits are as small as possible (not some arbitrarily high number) 🙏

> | max num accounts | 64 |

Reviewer 13: This is a good point. I think this SIMD should differentiate between sanitization limits and runtime limits

> | max num accounts | 64 |

Author: Added some additional information on runtime constants different than the current sanitization requirements for this transaction.

The `9LZdXeKGeBV6hRLdxS1rHbHoEUsKqesCC2ZAPTPKJAbK` is an odd one because technically if it had a SIMD it would just update this SIMD in the future. It's legacy so I'm not quite sure where to put it at this time or if we should just make a new SIMD for it now.

> | max num accounts | 64 |

Reviewer 13: Since we may increase the number of accounts allowed in a tx, we should make the max account constraint higher, preferably 96 as I explained in this [comment](https://github.com/solana-foundation/solana-improvement-documents/pull/296#issuecomment-3254187702)

> So the max we could raise the tx account limit without creating incentive to use ALT's is about (4096 - 1000) / 32 ~= 96 accounts

Otherwise we will need to inctroduce a new tx version to increase the max num accounts constraint

> | max num instructions | 64 |

Reviewer 13: This should probably be 255 for the sanitization limit, but we keep the runtime restriction of 64 total instructions (including CPI's). That way we can increase if desired in the future.

Reviewer 11: See https://github.com/solana-foundation/solana-improvement-documents/pull/296#discussion_r2321678823 - it would be great if we could keep these the same and later increase them both at the same time.

| max accounts/instruction | 128 |
| max UDP packets (fragmentation) | 6 |

> The above limits are useful to spec, but not necessarily different than what is

Reviewer 14: We should clarify what happens if any of these limits are violated?

Reviewer 5: Yes, any transaction violating these constraints is invalid and cannot be included in the chain. Violations are sanitization failures.

currently in place on the network.

### TransactionConfigMask

The transaction config mask is used to configure specific fee and resource 
requests in a transaction.
It is intended to be flexible enough to add new fields in the future, and is 
initially aimed at replacing the ComputeBudgetProgram instructions that are 
currently used to configure transactions.
Each bit in the mask represents 4-bytes in the `ConfigRequests` array.
If a configured value, such as priority-fee, needs more than 4-bytes a field can
use 2 bits in the mask.

Initially supported fields and assigned bits:

- [0, 1] - priority-fee. 8-byte LE u64.
- [2] - compute-unit-limit. 4 byte LE u32.
- [3] - requested loaded accounts data size limit. 4 byte LE u32.

> - [4] - requested heap size. 4 byte LE u32.

Reviewer 5: How we specify fees may change in the future, FD already has ideas on how they might want that to change. Similarly with these other limits. Unlikely we'll create enough to need 32 different values though

A new transaction format, `v1`, is proposed to enable larger transaction sizes.
The `v1` transaction format would be:

```
VersionByte (u8) - =129 to distinguish from legacy/v0 formats
LegacyHeader (u8, u8, u8) -- Required signatures from the current
`MessageHeader` type
NumInstructions (u8)
TransactionConfigMask (u32) -- Bitmask of which config requests are present.
LifetimeSpecifier [u8; 32]
NumAddresses (u8)
Addresses [[u8; 32]] -- Length matches NumAddresses
ConfigRequests [u8] -- Array of request values. (section size is popcount 
  TransactionConfigMask bytes). See section TransactionConfigMask for details.
InstructionHeaders [(u8, u8, u16)] -- Length of NumInstructions. Values are 
  (ProgramAccountIndex, NumInstructionAccounts, NumInstructionDataBytes)
InstructionPayloads [InstructionPayload] -- Length = NumInstructions.
  Each InstructionPayload is the concatenation of the following byte arrays:
    InstructionAccountIndexes [u8] -- Length = NumInstructionAccounts from the 
    corresponding InstructionHeader
    InstructionData [u8] -- Length = NumInstructionDataBytes from the 
    corresponding InstructionHeader
```

> The [TRAILING DATA SECTION] contains instruction's serialized account indexes

Reviewer 13: No longer needed right?

and instruction data bytes,
in the order they appear in the transaction. Each instruction's account indexes
must appear before its data bytes.
For example, if there are two instructions:

1. `Instruction { account_indexes: [0, 1], data: [2, 3, 4] }`
2. `Instruction { account_indexes: [5, 6], data: [7, 8] }`

The TRAILING DATA SECTION must look like this:

```
[0, 1, 2, 3, 4, 5, 6, 7, 8]
```

Lengths are serialized in the earlier `Ixs` section, and so are not repeated in
the trailing data section.

This new `v1` transaction format notably does not include address lookup
tables.

### Transaction Constraints

Any transaction violating these constraints will be considered invalid and will 
not be included in the chain. Violations are considered sanitization failures:

| max | value | new/existing |
| --- | --- | --- |
| max transaction size | 4096 | new |
| max num signatures per transaction | 42 | new |
| max num accounts | 96 | new |
| max num instructions | 64 | old |

> | max accounts/instruction | 255 | new |

Reviewer 13: this is old too

While the above lists sanitization failures, we will keep the current runtime 

> max limit of 12 signatures per transaction.

Reviewer 13: We actually don't have an explicit runtime max limit of 12 signatures. It's implicit right now. We should say that we are _adding_ a runtime max limit of 12 signatures to match the previous implicit limit from the previous max tx size.

We should also say that the runtime limit for the max number of tx accounts is unchanged from 64.

Author: Updated to account for this being an already implicit limit, as well as noting the current runtime limit for the max number of accounts

### TransactionConfigMask

The transaction config mask is used to configure specific fee and resource 
requests in a transaction.
It is intended to be flexible enough to add new fields in the future, and is 
initially aimed at replacing the ComputeBudgetProgram instructions that are 
currently used to configure transactions.
Each bit in the mask represents 4-bytes in the `ConfigRequests` array.
If a configured value, such as priority-fee, needs more than 4-bytes a field can
use 2 bits in the mask.

Initially supported fields and assigned bits:

- [0, 1] - priority-fee. 8-byte LE u64.
- [2] - compute-unit-limit. 4 byte LE u32.
- [3] - requested loaded accounts data size limit. 4 byte LE u32.
- [4] - requested heap size. 4 byte LE u32.

For 2 bit fields, such as priority-fee, both bits MUST be set. If only one of 
the bits is set, the transaction is invalid and cannot be included in blocks.

For TxV1 transactions, any ComputeBudgetProgram instructions are ignored for 
configuration, even if they are invalid.

> The instructions will still consume compute-units if included, like a successful
> no-op instruction.

Reviewer 13: ```suggestion
The instructions will still consume compute-units if included and be processed as a successful
no-op instruction.
```

does on the protocol level, namely that bundles are not guaranteed to be atomic
like singular transactions.

For these reasons, increasing the current transaction size limit would enable
developers to create new applications and use cases that were previously
unfeasible.

## New Terminology

- 'v1 transaction' - A new transaction format that is designed to enable larger
  transactions sizes while not having the address lookup table features 
  introduced in v0 transactions.

## Detailed Design

Increasing the transaction size limit above the current MTU max size is
possible with the use of QUIC. QUIC's RFC 9000 specification does not have an
explicit maximum stream size, allowing for larger transactions to be sent.

A new transaction format, `v1`, is proposed to enable larger transaction sizes.
The `v1` transaction format would be:

```
VersionByte (u8) - =129 to distinguish from legacy/v0 formats
LegacyHeader (u8, u8, u8) -- Required signatures from the current
`MessageHeader` type
NumInstructions (u8)
TransactionConfigMask (u32) -- Bitmask of which config requests are present.
LifetimeSpecificier [u8; 32]
NumAddresses (u8)
Addresses [[u8; 32]] -- Length matches NumAddresses

> ConfigRequests [u8] -- Array of request values. (section size is popcount 
>   TransactionConfigMask bytes). See section TransactionConfigMask for details.

Reviewer 13: I understand `TransactionConfigMask` explains how this works more but I think this comment should be updated to be a bit clearer. And I personally think "requests" is a strange name for this feature

```suggestion
ConfigValues [u32] -- Length equal to the popcount (number of set bits) of
  TransactionConfigMask. See section TransactionConfigMask for details.
```

Reviewer 13: Yeah each config value is 4 bytes. Could also do `[[u8; 4]]`

Reviewer 5: I'm not sure what the right "type" here is since it depends how many bits are associated with the value - some are u32 some are u64. All are some multiple of [u8; 4], so maybe that's the best?

> Ixs [(u8, u8, u16)] -- Number matches NumInstructions. Values are 
>  (program_account_index, num_accounts, num_data_bytes)
> [TRAILING DATA SECTION] -- Length must be consistent with the sum of 
>  num_accounts and num_data_bytes in `Ixs`

Reviewer 13: ```suggestion
InstructionHeaders [(u8, u8, u16)] -- Length of NumInstructions. Values are 
  (ProgramAccountIndex, NumInstructionAccounts, NumInstructionDataBytes)
InstructionPayloads [InstructionPayload] -- Length = NumInstructions.
  Each InstructionPayload is the concatenation of the following byte arrays:
    InstructionAccountIndexes [u8] -- Length = NumInstructionAccounts from the corresponding InstructionHeader
    InstructionData [u8] -- Length = NumInstructionDataBytes from the corresponding InstructionHeader```

Signatures [[u8; 64]] -- Length of `num_required_signatures` from
 `LegacyHeader`
```

The [TRAILING DATA SECTION] contains instruction's serialized account indexes
and instruction data bytes,
in the order they appear in the transaction. Each instruction's account indexes
must appear before its data bytes.
For example, if there are two instructions:

1. `Instruction { account_indexes: [0, 1], data: [2, 3, 4] }`
2. `Instruction { account_indexes: [5, 6], data: [7, 8] }`

The TRAILING DATA SECTION must look like this:

```
[0, 1, 2, 3, 4, 5, 6, 7, 8]
```

Lengths are serialized in the earlier `Ixs` section, and so are not repeated in
the trailing data section.

This new `v1` transaction format notably does not include address lookup
tables.

With the new transaction size limit, it is possible to go past some other 
implicit limitations that are currently in place on the network. To avoid
causing unintended issues, the following limitations are proposed to be in

> place:

Reviewer 13: I think the sanitization details and table deserve a new subheading of something like `Transaction Protocol Constraints`

The details here don't need to explain why we're doing this btw, it just needs to be declarative about how the validator should be implemented. So I would move your explanation about these constraints up under the new header and before the table since this is what an implementor needs to know.

> Any transaction violating these constraints will be considered invalid and will not be included in the chain. Violations are considered sanitization failures.

Author: Updated and moved around the section

| max | value |
| --- | --- |
| max transaction size | 4096 |
| max num signatures per transaction | 42 |
| max num accounts | 64 |
| max num instructions | 64 |
| max accounts/instruction | 255 |

> The above limits are useful to spec, but not necessarily different than what is

Reviewer 13: Can we just mark which constraints are new vs existing in the table? This section feels too vaguely worded.

Author: I believe I've updated all the correct new/existing tags. LMK if I got any wrong.

considered sanitization failures.

While the above lists sanitization failures, we will keep the current runtime 
max limit of 12 signatures per transaction.

### TransactionConfigMask

The transaction config mask is used to configure specific fee and resource 
requests in a transaction.
It is intended to be flexible enough to add new fields in the future, and is 
initially aimed at replacing the ComputeBudgetProgram instructions that are 
currently used to configure transactions.
Each bit in the mask represents 4-bytes in the `ConfigRequests` array.
If a configured value, such as priority-fee, needs more than 4-bytes a field can
use 2 bits in the mask.

Initially supported fields and assigned bits:

- [0, 1] - priority-fee. 8-byte LE u64.
- [2] - compute-unit-limit. 4 byte LE u32.
- [3] - requested loaded accounts data size limit. 4 byte LE u32.
- [4] - requested heap size. 4 byte LE u32.

For 2 bit fields, such as priority-fee, both bits MUST be set. If only one of 
the bits is set, the transaction is invalid and cannot be included in blocks.

For TxV1 transactions, any ComputeBudgetProgram instructions are ignored for 
configuration, even if they are invalid.

> The instructions will still consume compute-units if included.

Reviewer 13: Maybe say that they will still be processed as successful no-op instructions that consume compute units if included?

`MessageHeader` type
NumInstructions (u8)
TransactionConfigMask (u32) -- Bitmask of which config requests are present.
LifetimeSpecificier [u8; 32]
LifetimeSpecifier [u8; 32]
NumAddresses (u8)
Addresses [[u8; 32]] -- Length matches NumAddresses
ConfigRequests [u8] -- Array of request values. (section size is popcount 
  TransactionConfigMask bytes). See section TransactionConfigMask for details.
Ixs [(u8, u8, u16)] -- Number matches NumInstructions. Values are 
 (program_account_index, num_accounts, num_data_bytes)
[TRAILING DATA SECTION] -- Length must be consistent with the sum of 
 num_accounts and num_data_bytes in `Ixs`

> Signatures [[u8; 64]] -- Length of `num_required_signatures` from

Reviewer 2: Was this accidentally removed?

| max num signatures per transaction | 42 | new |
| max num accounts | 96 | new |
| max num instructions | 64 | old |

> | max accounts/instruction | 255 | old |

Reviewer 13: Sorry, I just checked and this is actually new. We currently only have a _runtime_ limit of 255 accounts/ix for top-level bpf program instructions and for any CPI instruction. There's no protocol constraint on accounts/ix so it's currently valid to pass over 255 accounts to a top-level builtin program instruction so this would be a new protocol constraint.

Author: Updated back

| max num signatures per transaction | 42 | new |
| max num accounts | 96 | new |
| max num instructions | 64 | old |
| max accounts/instruction | 255 | new |

While the above lists sanitization failures, we will add a runtime max limit of 
12 signatures per transaction to be consistent with the current implicit limit.
The runtime limit for the max number of accounts is not changed from the current
64, and the same goes for the max255 accounts per instruction also remaining 

> unchanged.

Reviewer 13: Technically the max 255 accounts per instruction is only for CPI's and top-level bpf program instructions. Top-level builtin program instructions don't have a max account limit. So there isn't a blanket runtime limit. I think we can leave this sentence out.

Author: Removed
